ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/sync/tasks/CompleteBlocksTaskTest.java

```diff
@@ -79,6 +95,115 @@ public void shouldCompleteWithoutPeersWhenAllBlocksAreEmpty() {
+    when(mockParisSpec.getWithdrawalsProcessor()).thenReturn(Optional.empty());
+    when(mockProtocolSchedule.getByBlockHeader((eq(header2)))).thenReturn(mockShanghaiSpec);
+    when(mockShanghaiSpec.getWithdrawalsProcessor())
+        .thenReturn(Optional.of(mockWithdrawalsProcessor));
+
+    final Block block1 =
+        new Block(header1, new BlockBody(Collections.emptyList(), Collections.emptyList()));
+    final Block block2 =
+        new Block(
+            header2,
+            new BlockBody(
+                Collections.emptyList(),
+                Collections.emptyList(),
+                Optional.of(Collections.emptyList())));
+
+    final List<Block> expectedBlocks = asList(block1, block2);
+    final EthTask<List<Block>> task =
+        CompleteBlocksTask.forHeaders(
+            mockProtocolSchedule,
+            ethContext,
+            List.of(header1, header2),
+            maxRetries,
+            new NoOpMetricsSystem());
+    assertThat(task.run()).isCompletedWithValue(expectedBlocks);
+  }
+
+  @Test
+  public void shouldCompleteWhenBlocksAreEmptyExceptForWithdrawals() {
+    final ProtocolSchedule mockProtocolSchedule = Mockito.mock(ProtocolSchedule.class);
+    final ProtocolSpec mockParisSpec = Mockito.mock(ProtocolSpec.class);
+    final ProtocolSpec mockShanghaiSpec = Mockito.mock(ProtocolSpec.class);
+    final WithdrawalsProcessor mockWithdrawalsProcessor = Mockito.mock(WithdrawalsProcessor.class);
+
+    final Withdrawal withdrawal =
+        new Withdrawal(UInt64.ONE, UInt64.ONE, Address.fromHexString("0x1"), GWei.ONE);
+    final List<Withdrawal> withdrawals = List.of(withdrawal);
+    final Hash withdrawalsRoot = BodyValidation.withdrawalsRoot(withdrawals);
+
+    final BlockHeader preWithdrawalsHeader = new BlockHeaderTestFixture().number(1).buildHeader();
+    final BlockHeader firstWithdrawalsHeader =
+        new BlockHeaderTestFixture().number(2).withdrawalsRoot(withdrawalsRoot).buildHeader();
+    final BlockHeader secondWithdrawalsHeader =
+        new BlockHeaderTestFixture().number(3).withdrawalsRoot(Hash.EMPTY_TRIE_HASH).buildHeader();
+
+    final Block block1 = new Block(preWithdrawalsHeader, BlockBody.empty());
+    final Block block2 =
+        new Block(
+            firstWithdrawalsHeader,
+            new BlockBody(
+                Collections.emptyList(), Collections.emptyList(), Optional.of(withdrawals)));
+    final Block block3 =
+        new Block(
+            secondWithdrawalsHeader,
+            new BlockBody(
+                Collections.emptyList(),
+                Collections.emptyList(),
+                Optional.of(Collections.emptyList())));
+    final List<Block> expected = asList(block1, block2, block3);
+
+    final RespondingEthPeer respondingPeer =
+        EthProtocolManagerTestUtil.createPeer(ethProtocolManager, 1000);
+    final RespondingEthPeer.Responder responder = responderForFakeBlocks(expected);
```

Comment:
Why does the responder need all the blocks? It should only need to download block2 as the others are empty

Reply:
it doesn't we thought it was a good test - e.g. let's say the code under test had a bug where it requested 3 blocks instead of 1, but we only ever returned one block in the responser, we wouldn't catch that.

Reply:
It does not, you're right, good catch!